File: src/functions/incidentPassthrough.py
#!/usr/bin/env python3
"""
Pulsar Function that returns the input message. Used for decoupling flows and creating stop valves.
"""
import pulsar
from src.common.incident import Incident
from pulsar.schema import AvroSchema

class Schema:
    schema = None

    def __init__(self, *args):
        self.schema = args[0]

    def __call__(self, f):
        def wrapped(*args):
            args = list(args)
            args[1] = self.schema.decode(args[1].encode())
            return self.schema.encode(f(*tuple(args))).decode("utf-8")

        return wrapped

class IncidentPassthrough(pulsar.Function):
    def __init__(self):
        self.userConfig = None
        self.logger = None
        self.isInitialized = False
    
    def initialize(self, context):
        self.logger = context.get_logger()
        self.userConfig = context.user_config()
        self.isInitialized = True
    
    @Schema(AvroSchema(Incident))
    def process(self, message, context):
        if self.isInitialized == False:
            self.initialize(context)
        return message

File: java/incident-passthrough/src/main/java/com/trafficcorp/example/demofunctions/function/IncidentPassthrough.java
package com.trafficcorp.example.demofunctions.function;

import org.apache.pulsar.functions.api.Context;
import org.apache.pulsar.functions.api.Function;

public class IncidentPassthrough implements Function<Incident, Incident> {
    
    public IncidentPassthrough() {
        
    }

    @Override
    public Incident process(Incident input, Context context) throws Exception {
        return input;
    }
}